token/transfer-hook/cli/src/meta.rs

```diff
@@ -0,0 +1,318 @@
+use {
+    serde::{Deserialize, Serialize},
+    solana_sdk::pubkey::Pubkey,
+    spl_tlv_account_resolution::{account::ExtraAccountMeta, seeds::Seed},
+    std::{fmt, path::Path, str::FromStr},
+    strum_macros::{EnumString, IntoStaticStr},
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, EnumString, IntoStaticStr, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+#[strum(serialize_all = "kebab-case")]
+enum AccountMetaRole {
+    Readonly,
+    Writable,
+    ReadonlySigner,
+    WritableSigner,
+}
+impl AccountMetaRole {
+    fn bools(&self) -> (/* is_signer */ bool, /* is_writable */ bool) {
+        match self {
+            AccountMetaRole::Readonly => (false, false),
+            AccountMetaRole::Writable => (false, true),
+            AccountMetaRole::ReadonlySigner => (true, false),
+            AccountMetaRole::WritableSigner => (true, true),
+        }
+    }
+}
+impl fmt::Display for AccountMetaRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct ExtraAccountMetaConfigFile {
+    #[serde(alias = "extraMetas")]
+    extra_metas: Vec<ExtraAccountMetaConfig>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct ExtraAccountMetaConfig {
+    pubkey: Option<String>,
+    seeds: Option<Vec<Seed>>,
+    role: Option<AccountMetaRole>,
+    #[serde(alias = "isSigner")]
+    is_signer: Option<bool>,
+    #[serde(alias = "isWritable")]
+    is_writable: Option<bool>,
+}
+impl ExtraAccountMetaConfig {
+    fn discriminator_and_address(&self) -> (u8, [u8; 32]) {
+        if self.pubkey.is_some() && self.seeds.is_some() {
+            panic!("Only one of `pubkey` or `seeds` must be present");
+        }
+        if let Some(pubkey_string) = &self.pubkey {
+            (0, Pubkey::from_str(pubkey_string).unwrap().to_bytes())
+        } else if let Some(seeds) = &self.seeds {
+            (1, Seed::pack_into_address_config(seeds).unwrap())
```

Comment:
Now that the discriminants are being used outside of the crate, I think it's time for a separate enum in spl-tlv-account-resolution

Reply:
Ehh, this may be more trouble that it's worth, since you can't derive `Pod` on an enum, and the discriminator for external PDAs is not static.

Let me instead try to refactor this code.

Reply:
Yeah, I've got a better solution. Reminds me I also need to add support for external PDAs.